### DISCUSSION TITLE
building for production with a custom server

### DISCUSSION THREAD

**Discussion by samuelstroschein** [vikejs/vike#560] [Discussion] [answered]
I am trying to build a production build with VPS and express. The boilerplate provides [an example](https://github.com/brillout/vite-plugin-ssr/blob/main/boilerplates/boilerplate-react-ts/server/index.ts) that differs from vite's build for production guide. The latter does not work.

## Context

The boilerplate uses `ts-node` to serve the production build. I have some polyfill issues that I am trying to solve by using vite to bundle my server and run a simple `node ./dist/server/main.js`. Regardless of ts-node being an issue, I expected vite's build for production to work with VPS.

## Problem

Building for production as provided in [vite's documentation](https://vitejs.dev/guide/ssr.html#building-for-production) does not work. Vite recommends two build scripts, one for the client and one for the server.

```ts
{
  "scripts": {
    "dev": "node server",
    "build:client": "vite build --outDir dist/client",
    "build:server": "vite build --outDir dist/server --ssr src/entry-server.js"
  }
}
```

Defining the build:server script fails with VPS and `disableAutoFullBuild ` set to false fails though:

```
[@brillout/vite-plugin-import-build:vite-plugin-ssr] `dist/server/pageFiles.js` is missing: make sure your Rollup config doesn't change the name of the file `pageFiles.js` 
```

The same problem applies to `telefunc`, if telefunc is passed into the plugins: 

```
[telefunc:buildConfig:assert] [telefunc@0.1.40][Wrong Usage] You seem to be using a tool that conflicts with Telefunc. Reach out to a Telefunc maintainer. (Info for maintainer: couldn't find Telefunc's Rollup input entry.)
```

## Questions

1. Should building for production according to vite's docs work with VPS in the first place? 
2. How does the boilerplate example `serve` work? 
Does vps `render` look up the files in `dist/client/*` via `serve`? 
What happens with all the files in `dist/server/*`?

**Comment by brillout** [comment, marked as answer]
You may want to give https://github.com/cyco130/vavite a try.

Also: https://github.com/hattipjs/hattip/tree/main/examples/vavite/vite-plugin-ssr. HatTip is a more edgy, but really cool :-).

FYI https://vite-plugin-ssr.com/disableAutoFullBuild.

**Reply by samuelstroschein** [reply]
`disableAutoFullBuild` does not work and leads to the stated problem. I'll follow up with a repro tomorrow

**Reply by brillout** [reply]
It's probably due to `--outDir dist/server`. Also using VPS with a server entry `--ssr src/entry-server.js` is something I never tried.

**Reply by brillout** [reply]
I did a bit of digging. So `--outDir dist/server` isn't the problem. The culprit is `--ssr src/entry-server.js`. It seems like Vite is erasing all other entries. So it's more of a Vite bug. It's a bummer, because ` --ssr src/entry-server.js` is actually a good idea (I could have made it work without the need to set `disableAutoFullBuild`).

**Reply by samuelstroschein** [reply]
That is a bummer. The `--ssr` flag seemed like a natural way to provide the entry point for the server. 

I struggle to understand what happens when vite-plugin-ssr is used with serv. Is the compiled output of `/dist/server` used at all in the boilerplate examples?

**Reply by samuelstroschein** [reply]
Building for production is .... complicated. 

Setting `disableAutoFullBuild` to `true` leads to the VPS and telefunc crashes during the build step "can't find dist/server/*".  

Setting `disableAutoFullBuild` to `false` and using vite's `--ssr` flag after the client (aka full build has finished) leads to 
```
[vite-plugin-ssr@0.4.46][Warning] The CLI call `$ vite build --ssr` is superfluous since `$ vite build` also builds the server-side. If you want two separate build steps then use https://vite-plugin-ssr.com/disableAutoFullBuild or use Vite's `build()` API.
```

There seems to be no none-hacky way to bundle the server entry with vite and vite-plugin-ssr

**Reply by brillout** [reply]
> I struggle to understand what happens when vite-plugin-ssr is used with serv. Is the compiled output of /dist/server used at all in the boilerplate examples?

https://github.com/brillout/vite-plugin-ssr/issues/403

> There seems to be no none-hacky way to bundle the server entry with vite and vite-plugin-ssr

Yes. The recommendation so far is to use vavite or HatTip. I agree a simpler solution would be nice. I just created a ticket for that https://github.com/brillout/vite-plugin-ssr/issues/562.

> Building for production is .... complicated.

I agree it should be easier.
